### PR TITLE
Updated Spotlight Examples

### PR DESCRIPTION
**Pull request by MasterJames**
This optimizes the examples for mobile and general performance optimization of angle to shadow map size.
I was going to use the Stats FPS but had trouble accessing it and made a PR there https://github.com/mrdoob/stats.js/pull/53 for that, but realized I needed to build a dedicated FPS.js solution for a goal of making the initial example not embarrassing on mobile (as it in fact crashes). The issue is map size and on my dev system I get 120 frames with 8192 map size with 3 lights, so no problems revealed themselves, but on mobile it does crash etc.

This new method adjusts the map size based on the angle (an interesting compute to see in action in the gui).
Further in the animated multiple spotlights example a Frames Per Second average relative to the desired FPS diminishes the automatically calculated map size further, in an attempt to achieve that speed.

So if the FPS.js could move to core or something in the future that would be great.

If beloved we would then integrate somehow the automatic map size setting function into the spotlight angle update stuff directly.

I know it's a fairly intense (but needed) patch to the original examples but it's got long term purpose too. This is something I hinted at months ago, but is the perfect time to introduce these kinds of FPS optimizations into things. Another example might be a distance based factor that diminishes texture size for objects further away.

I'll probably post some of the other interesting issues revealed in separate comments later.

Basically it's nice to see these things in motion and adjustable as it can reveal more issues.

### CODE REVIEW

**Comment by MasterJames** [comment]
This new FPS is currently designed and demonstrated using the option to pass an object into it to store it's data in. Then when it update()s it changes the data in whatever targetObject you want. If not it hides an internal one with the traditional function getters. for fps() and average() which take a precision so they are slightly different. and one to get the data object too.
When you update you can suppress the compute and then when you need it call the fps() function, which will compute it (average() works similarly).
Average is computed in an intelligent way. If you jump from 120 FPS to 1 FPS it looks at less of it's buffer to compute the average from so average is over about 3 seconds regardless of the frame rate.
The fps log array length is adjustable but the default works well in most situations I figure.

I imagine either integrating an event listening system for FPS thresholds to trigger events or functions etc. or maybe with this pass-the-data-object-in design, one could add listeners or watches as needed in there own code. Maybe a subclass that adds events is a better route? Well at least for some options.

**Comment by MasterJames** [comment]
@bhouston I figured you're the best to answer this question I discovered with this PR.
Thanks to the examples we can now see in action all the setting and map sizes etc. which could reveal some hidden issues like this one I think demonstrates.
If I put as in this example [which demonstrates this issue nicely] (or any spotlight) an initial shadow map size, it will persist.
My question is "How do I reset that?", because if you set the shadow map size low initially and then zoom the map size up it looks blocky still, nice an crisp but from the initial size. [Obviously if you set it big initially and then shrink it down as an optimization it leaves a larger map in the cache or something.
I guess the question is, is that the best way as is and not to worry about it or maybe it's a bug this demonstrates, or maybe it needs a special ability to rest the initial map data in this case with dynamic shadow map size being set by angle.

I'm also wondering if animating the map size with the angle, while if it's also potentially resting the map data cache, if it would be good, bad, and/or ugly. [so maybe on-anim-complete it resets the map data].
Anyway if it doesn't animate the map size, an extreme change could potential pop the shadow map size abruptly.

**Comment by mrdoob** [comment]
Hmm, I think the FPS code is out of scope for this... 😕

Also, `webgl_lights_spotlights` consistently freezes my machine and forces me to reboot ☹️

**Comment by mrdoob** [comment]
Specially sad because I was downloading a 20GB file the whole morning and I was almost done...

**Comment by bhouston** [comment]
Can you @MasterJames separate out the angle-based map size changes from the FPS changes?  The angle-based map size changes is nearly exactly to how Unity 3D works.

**Comment by bhouston** [comment]
There is something in those spotlight examples that is causing issues for I think everyone.  I've got a top end Window PC (e.g. recent Intel i7 Extreme, GeForce 980 /w 4GB of ram, etc.) and it gets horrible FPS (drops to like ~3 fps) on these examples if I play around with them.

**Comment by WestLangley** [comment]
>Also, webgl_lights_spotlights consistently freezes my machine and forces me to reboot

https://github.com/mrdoob/three.js/pull/8485#issuecomment-204475469

**Comment by MasterJames** [comment]
Oh wow sorry guys. Maybe I should have used requestAnimationFrame but it didn't function as well as using setTimeout to maintain a desired frame rate. I guess I will have to purge the frame speed monitoring as soon as possible.

It could also be because I set the size to a max of 8192 and then scale down with automation or it looks blocky as explained in question to Ben. It shouldn't takeout the whole machine though.

Maybe it's okay with initial size reduced? but again blockiness results when scaling up from a lower initial setting. Is there a way to reset the map size's buffer after it initially loads?

I did notice if on mobile that if the frame rate was low it could catch up to it's maximum ability by raising the desired FPS, but if it exceeded it it would just hang and get a super low frame speed, as if it was rendering but had no time to update the screen buffer. Then maybe it backs up and crashes. It then seemed even more important to consider ways to optimize the frame speed.

Thanks for the feedback I will test this through other machines first as well before I submit any fixes.
Maybe I'll be able to reproduce the crashes locally and get a handle on what caused that, otherwise the plan is to just stick with the angle optimization.

**Comment by MasterJames** [comment]
I patched this temporarily with requestAnimationFrame, which stops the runaway freeze-up/crash.

`					window.requestAnimationFrame( render );`

but the PR has not been adjusted properly yet.

**Comment by MasterJames** [comment]
Okay @mrdoob Mega apologies again, but this is safe to surf now!, **but please purge your cache first.**

It seems I was only able (with a boundless computer system) to render 120 FPS without resorting to calling render from a setTimeout, which caused a catch-up/runaway problem (aka crash). 
Using requestAnimationFrame my system is limited to 60 FPS and when set to it it also falls short without an adjustment to compensation. It also never achieved the desired frame rate so I did call render from a setTimeout which in hindsight was a pretty junior move but again it was accurate and twice as fast, albeit on a system that has no issues with 120 FPS and  map sizes of 8192, so hopefully you can understand how and why I ended up with that dangerous solution.

It is also noted that I could not achieve a super nice shadow edge without getting up past 4096, but it was configured to automatically scale to it's limits nicely with 8192. Meaning this is visually when with automatic scaling it looks good and believable below that... not so. Again why I was drawn to set that as an initial maximum on start up because they don't update map size changes for some reason I'm asking Ben about later below.

I removed the FPS stuff after getting it working fine and committed a commented out version for the record. Even though it was in fact working well and super simple addition.
What this revealed on another test machine with win7 that only gets 1 FPS is that reducing the map size when the desired FPS was not being obtained did not result in an increase in FPS, so in this case it was not needed. Originally I had no way to know testing on the system without limits. 

[This hardware was almost top of the line a few years back, so not uncommon for gamers these days I suspect.]

```
NVIDIA GeForce GTX 780 Ti
	Manufacturer	NVIDIA
	Model	GeForce GTX 780 Ti
	Device ID	10DE-100A
	Revision	A2
	Subvendor	ASUStek Computer Inc (1043)
	Current Performance Level	Level 1
	Current GPU Clock	324 MHz
	Current Memory Clock	324 MHz
	Voltage	0.887 V
	Bus Interface	PCI Express x16
	Temperature	38 °C
	Driver version	9.18.13.4709
	BIOS Version	80.80.30.00.01
	Physical Memory	3071 MB
	Virtual Memory	3072 MB
		Count of performance levels : 4
				Level 1 - "2D Desktop"
					GPU Clock	324 MHz
					Memory Clock	324 MHz
				Level 2 - "Default"
					GPU Clock	536 MHz
					Memory Clock	810 MHz
				Level 3 - "3D Applications"
					GPU Clock	549 MHz
					Memory Clock	3500 MHz
				Level 4 - "3D Applications"
					GPU Clock	549 MHz
					Memory Clock	3500 MHz

```

Anyway I've tested it on a old system (1FPS), and a sluggish laptop as well, and it simple doesn't go nuts and crash with RequestAnimationFrame of course.

@bhouston  Hopefully now we can see the related issue via the example I still ponder that I'm asking for you thoughts about please and thank you.

If you set the initial max map size (in the code manually) to a small vs big value, it keeps it's original resolution and so if you zoom the angle to 1.5 radians which is wide open you see blocky low res pixel shapes with high definition, while if you start with a super high res it zooms dynamically up with loads more full detail and looks nicer and not blocky (like it looks if the maps size is initially low).

So I was wondering if when the map size is changed it could rebuild that cached data or what the function to update it properly would be. Then maybe we would see and improvement if FPS when lowering and "optimizing" them, but otherwise it appears no noticeable difference will be likely achieved in FPS by lower map sizes.

So if it makes no difference and you need the max map size you'll ever possibly scale the spotlight's angle to initially, then you'd want to be able to inspect the available RAM on the GPU or something to be able to properly provide an optimized map size. Especially when the initial example with a 8192 map size was crashing some smart phones browsers, likely due to GPU RAM overflow.
Would there be some kind of a try-catch block on the renderer.render call solution to catch this from happening/crashing from exceeding the GPU's RAM possibly? already? events?

_Anyway I hope you find this version is the best and most stable way so far to demonstrate the wonders of the mighty spotlights!_
Reminder: The first "merged" PR of these Examples had a mostly mobile issues due to the map size being set to 8192.

**Comment by WestLangley** [comment]
You are not changing the shadow map size by simply resetting `shadow.mapSize`,  as the render target on the GPU is left unchanged. You need to use this pattern.
```javascript
light.shadow.mapSize.width = 256;
light.shadow.mapSize.height = 256;
light.shadow.map.dispose();
light.shadow.map = null;
```
Also, I prefer simpler examples, to be honest.

**Comment by MasterJames** [comment]
Thanks for the dispose method. I wondered if it seems maybe this is something similar to the changes in angle that it should be automatically done when setting mapSize? for now I'm refactoring this into the example, thanks. I suppose one might want to not do it automatically, but then again maybe there's really no use-case not to have the map disposed of automatically.

**Comment by MasterJames** [comment]
Okay it's ready! for a second round. These both have an added GL error detection system that pops an alert warning you it's already adjusted to compensate. So no way to crash it now. In theory. Note that Mobile also is set to half of the default max map size.
Having halved the original submissions default max map size, I noted on high-end systems the automation sized shadows look blocky and so I added a switch to double the max map sizes, which is really as it's intended to function (and doesn't go high when angles are more reasonable), and to my eye when it is passable without penumbra blurring (so for a crisp enough edged on the shadows of an extreme angle the shadow map looks good still, when it's normal automation keeps it low). 
The function to scale is designed to be tardy relative that of a one to one ratio scale up, but catches up in the end, and also has a minimum of 256.
This "doubling" seems to not only allow for a consistent experience through an animated spotlight angle change, but pretty much requires it to look right. Still for slower systems (and lower yet again on mobile) the examples have been retarded in this way so to hopefully not perform differently on these varied devices.

Anyway we lost the FPS system because for shadow map sizes it made no difference in performance. Map size should probably be calculated based on some larger understanding of your total texture size. Still this automation system set to a lowest (max) for visually acceptable/tolerable level of detail in shadows, demonstrates how scalable dynamics can maximize your performance in animated dynamic interactive situations and allow for more texture memory when and where you need it (like 'camera distance', which is not demonstrated in this example).
_
So again, one the solution is smoother looking when animated, and two optimizing your texture map sizes aka memory footprint on the fly. With added corrective ability for errors and limits._

Without monitoring FPS this idea is only half baked in this demonstration, because the goal is to optimize that so you want to have a good average fps it adjust with, but use an automated system of priorities to maximize FPS and form & function etc. Anyway it demonstrates one simple solution.
So if we add that option to automatically adjust the spotlight in the spotlights code?! It would be like we just were doing with the setters/getters type approach for 'angle' to shadow camera FOV which @mrdoob brilliantly interpreted recently. We should also simply add that there in a similar fashion so people don't have to think about it too much (just an weighted set of preferences I imagined). 

Still, if we're just talking about automation with only shadow map size, this makes shadows that are animating their angles visually acceptable, and hassle free.

Extra: dat GUI that adjust it's positioning and is different on mobile, so again it feels consistent across multiple user experiences.

**Comment by MasterJames** [comment]
Okay still learning how to dance the Git Hubabuba.
This now rebased has resolved it's conflicts.
This is the Most Stable version, the currently merged one will crash on mobile this fixes all that.

**Comment by MasterJames** [comment]
@tschw and/or @WestLangley could you check this over and give a thumbs up to @mrdoob for it to be merged? [Pretty Please.] This initial patch had crashed his system and things are crazy busy right now as you well have noticed, so it's I believe "ready to Go!".
It's important because the initial already Merged PR for the new SpotLight Examples needs proper handling for Mobile, as the previous merged one potentially generates a GL errors when a shadow map maximum set to 8K is too much for the running system. Sooo embarrassing. As my first example I've learned now to double check it on multiple systems as well as mobile, which has been done with it's current state now (last week).
Otherwise I would need some further guidance on getting it on course with the high quality we all expect from THREE, thank you.

**Comment by mrdoob** [comment]
> It's important because the initial already Merged PR for the new SpotLight Examples needs proper handling for Mobile, as the previous merged one potentially generates a GL errors when a shadow map maximum set to 8K is too much for the running system.

I change this some weeks ago already. The shadows now use a 1024 texture...
https://github.com/mrdoob/three.js/blob/master/examples/webgl_lights_spotlight.html#L92-L93
https://github.com/mrdoob/three.js/blob/master/examples/webgl_lights_spotlights.html#L137-L138

**Comment by MasterJames** [comment]
Oh I missed you comment @mrdoob somehow?! I was just stopping in to ask about this. 
Can you take a minute to consider this PR? It handles things much better now.
Otherwise if you could explain how it's not desirable or suggestions to improve it so it is, I would appreciate the extra time for your feedback, thanks.

Also note the other original version doesn't reset the shadow map size properly like this one does thanks to advice given since as well. I think that's an easy patch otherwise too.

Also I've been reading every email/issue for about 6 months and will be travelling for business and personal affairs too over the next few months so unfortunately my participation and attempt to learn through monitoring the activity here will be on hold until August, but I will check back for your (final) response to decide if this will be closed or merged. Again thanks, I've noticed it's been particularly busy the last few weeks, so no rush on this if you've patched the map size already.
